src/marlin_stubs/G26.cpp

```diff
@@ -0,0 +1,173 @@
+
+static const constexpr float pi = 3.1415926535897932384626433832795f;
+
+class FirstLayer : public FSM_Holder {
+private:
+    uint16_t total_lines = 1;
+    uint16_t current_line = 0;
+
+public:
+    FirstLayer()
+        : FSM_Holder(ClientFSM::FirstLayer, 0) {}
+
+    void wait_for_move() {
+        planner.synchronize();
+    }
+
+    /// Moves head and extrudes
+    /// Use NAN for axis you don't want to move
+    /// \param e is relative extrusion
+    /// \param f is defined in millimeters per minute
+    void go_to_destination(const float x, const float y, const float z, const float e, const float f) {
+        if (isfinite(x))
+            destination[0] = x;
+        else
+            destination[0] = current_position[0];
+
+        if (isfinite(y))
+            destination[1] = y;
+        else
+            destination[1] = current_position[1];
+
+        if (isfinite(z))
+            destination[2] = z;
+        else
+            destination[2] = current_position[2];
+
+        if (isfinite(e))
+            destination[3] = current_position[3] + e;
+        else
+            destination[3] = current_position[3];
+
+        if (isfinite(f))
+            feedrate_mm_s = f / 60.f;
+
+        prepare_move_to_destination();
+    }
+
+    /// increases progress by 1 line and sends it to Marlin
+    void inc_progress() {
+        current_line++;
+        const variant8_t var = variant8_i8(current_line / (float)total_lines);
+        marlin_set_var(MARLIN_VAR_SD_PDONE, var);
+    }
+
+    void go_to_destination_and_wait(const float x, const float y, const float z, const float e, const float f) {
+        go_to_destination(x, y, z, e, f);
+        wait_for_move();
+        inc_progress();
+    }
+
+    /// Moves and extrudes
+    /// Keep Z and feedrate from the last time
+    void go_to_destination(const float x, const float y, const float e) {
+        go_to_destination(x, y, NAN, e, NAN);
+    }
+
+    /// @returns length of filament to extrude
+    float extrusion(const float x1, const float y1, const float x2, const float y2, const float layerHeight = 0.2f, const float threadWidth = 0.5f) {
+        const float length = sqrt(SQR(x2 - x1) + SQR(y2 - y1));
+        return length * layerHeight * threadWidth / (pi * SQR(filamentD / 2));
+    }
+
+    float extrusion_Manhattan(const float *path, const uint32_t position, const float last) {
+        if (position % 2 == 0) {
+            const float x = path[position];
+            const float y = path[position - 2];
+            return extrusion(x, y, last, y, layerHeight, threadWidth);
+        } else {
+            const float x = path[position - 1];
+            const float y = path[position];
+            return extrusion(x, y, x, last, layerHeight, threadWidth);
+        }
+    }
+
+    void print_snake(const float *snake, const size_t snake_size, const float speed) {
+
+        /// move to start
+        go_to_destination(snake[0], snake[1], 0); // Process X Y Z E F parameters
```

Comment:
this assumes snake has at least 2 items ... a static_assert or runtime check would be appropriate to add